File: Player/src/main/java/ru/testsimpleapps/coloraudioplayer/ui/adapters/PlaylistAdapter.java
package ru.testsimpleapps.coloraudioplayer.ui.adapters;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import ru.testsimpleapps.coloraudioplayer.R;
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.IPlaylist;
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.cursor.CursorFactory;
import ru.testsimpleapps.coloraudioplayer.managers.tools.CursorTool;
import ru.testsimpleapps.coloraudioplayer.managers.tools.TimeTool;


public class PlaylistAdapter extends BaseAdapter {

    private final Context mContext;
    private IPlaylist mIPlaylist;
    private boolean mIsExpand = true;
    private long mPreviousSearchPosition = 0;
    private long mSearchedPosition = IPlaylist.ERROR_CODE;
    private int mViewPadding;

    public PlaylistAdapter(@NonNull Context context) {
        mContext = context;
        mViewPadding = (int)context.getResources().getDimension(R.dimen.playlist_item_selection_padding);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        switch (i) {
            case TYPE_HEADER: {
                final View viewItem = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.playlist_header, viewGroup, false);
                return new ViewHolderHeader(viewItem);
            }

            case TYPE_ITEM: {
                final View viewItem = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.playlist_item_layout, viewGroup, false);
                return new ViewHolderItem(viewItem);
            }

            default:
                return null;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
        if (mIPlaylist != null) {
            if (viewHolder instanceof ViewHolderItem) {
                final ViewHolderItem mViewHolder = (ViewHolderItem) viewHolder;
                mIPlaylist.goToPosition(i - 1);

                // Main
                mViewHolder.mImageTrack.setImageResource(R.drawable.image_item_track);
                mViewHolder.mCountTrack.setText(String.valueOf(i));
                mViewHolder.mNameTrack.setText(String.valueOf(mIPlaylist.getTrackName()));

                // Info
                if (mIsExpand) {
                    mViewHolder.mInfoLayout.setVisibility(View.VISIBLE);
                    mViewHolder.mDurationTrack.setText(TimeTool.getDuration(mIPlaylist.getTrackDuration()));
                    mViewHolder.mArtistsTrack.setText(String.valueOf(mIPlaylist.getTrackArtist()));
                    mViewHolder.mAlbumTrack.setText(String.valueOf(mIPlaylist.getTrackAlbum()));
                    mViewHolder.mDateTrack.setText(TimeTool.getDateTime(mIPlaylist.getTrackDateModified()));
                } else {
                    mViewHolder.mInfoLayout.setVisibility(View.GONE);
                }

                // Background for current position or image_search
                if (mIPlaylist.getTrackId() == CursorFactory.getInstance().getTrackId()) {
                    mViewHolder.itemView.setBackgroundResource(R.drawable.drawable_recycleview_item_selection);
                    mViewHolder.itemView.setPadding(mViewPadding, mViewPadding, mViewPadding, mViewPadding);
                } else if (mSearchedPosition == i) {
                    mViewHolder.itemView.setBackgroundResource(R.drawable.drawable_recycleview_item_find);
                    mViewHolder.itemView.setPadding(mViewPadding, mViewPadding, mViewPadding, mViewPadding);
                } else {
                    mViewHolder.itemView.setBackgroundColor(Color.TRANSPARENT);
                    mViewHolder.itemView.setPadding(0, 0, 0, 0);
                }

            } else if (viewHolder instanceof ViewHolderHeader) {
                final ViewHolderHeader mViewHolder = (ViewHolderHeader) viewHolder;
                mViewHolder.mNamePlaylist.setText(CursorTool.getPlaylistNameById(mContext.getContentResolver(),
                        mIPlaylist.getPlaylistId()));
                mViewHolder.mCountTracks.setText(String.valueOf(getTotalTracks()));
                mViewHolder.mTotalTimeTracks.setText(TimeTool.getDuration(getTotalTime()));
            }
        }
    }

    @Override
    public int getItemCount() {
        return (mIPlaylist != null ? (int) mIPlaylist.size() : 0) + 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return TYPE_HEADER;
        }
        return TYPE_ITEM;
    }

    @Override
    public long getItemId(int position) {
        if (mIPlaylist != null && mIPlaylist.goToPosition(position)) {
            return mIPlaylist.getTrackId();
        }

        return IPlaylist.ERROR_CODE;
    }

    public void setPlaylist(final IPlaylist iPlaylist) {
        mIPlaylist = iPlaylist;
        notifyDataSetChanged();
    }

    public void setExpand(final boolean isExpand) {
        mIsExpand = isExpand;
        notifyDataSetChanged();
    }

    public int searchMatch(final String text) {
        if (mIPlaylist != null) {
            final long position = mIPlaylist.find(mPreviousSearchPosition, text);
            mPreviousSearchPosition = position + 1;
            return (int)position;
        }

        return (int)IPlaylist.ERROR_CODE;
    }

    public void setSearchedPosition(final long position) {
        mSearchedPosition = position;
    }

    private long getTotalTime() {
        return mIPlaylist != null? mIPlaylist.getTotalTime() : 0;
    }

    private long getTotalTracks() {
        return mIPlaylist != null? mIPlaylist.size() : 0;
    }

    protected class ViewHolderHeader extends RecyclerView.ViewHolder {
        @BindView(R.id.playlist_header_name_value)
        TextView mNamePlaylist;
        @BindView(R.id.playlist_header_count_value)
        TextView mCountTracks;
        @BindView(R.id.playlist_header_total_time_value)
        TextView mTotalTimeTracks;

        public ViewHolderHeader(final View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    protected class ViewHolderItem extends RecyclerView.ViewHolder {

        /*
        * Main views
        * */
        @BindView(R.id.playlist_track_image)
        ImageView mImageTrack;
        @BindView(R.id.playlist_track_count)
        TextView mCountTrack;
        @BindView(R.id.playlist_track_name)
        TextView mNameTrack;

        /*
        * Optional
        * */
        @BindView(R.id.playlist_track_item_info_layout)
        ConstraintLayout mInfoLayout;
        @BindView(R.id.playlist_track_duration_value)
        TextView mDurationTrack;
        @BindView(R.id.playlist_track_artist_value)
        TextView mArtistsTrack;
        @BindView(R.id.playlist_track_albums_value)
        TextView mAlbumTrack;
        @BindView(R.id.playlist_track_date_value)
        TextView mDateTrack;

        public ViewHolderItem(final View view) {
            super(view);
            ButterKnife.bind(this, view);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onItemClick(view, getLayoutPosition() - 1);
                    }
                }
            });
        }
    }

}

File: player/src/main/java/ru/testsimpleapps/coloraudioplayer/ui/adapters/PlaylistAdapter.kt
package ru.testsimpleapps.coloraudioplayer.ui.adapters

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.support.constraint.ConstraintLayout
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import butterknife.BindView
import butterknife.ButterKnife
import ru.testsimpleapps.coloraudioplayer.R
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.IPlaylist
import ru.testsimpleapps.coloraudioplayer.managers.player.playlist.cursor.CursorFactory
import ru.testsimpleapps.coloraudioplayer.managers.tools.CursorTool
import ru.testsimpleapps.coloraudioplayer.managers.tools.TimeTool


class PlaylistAdapter(private val mContext: Context) : BaseAdapter() {

    private var mIPlaylist: IPlaylist? = null
    private var mIsExpand = true
    private var mPreviousSearchPosition: Long = 0
    private var mSearchedPosition = IPlaylist.ERROR_CODE
    private val mViewPadding: Int

    private val totalTime: Long
        get() = if (mIPlaylist != null) mIPlaylist!!.totalTime else 0

    private val totalTracks: Long
        get() = if (mIPlaylist != null) mIPlaylist!!.size() else 0

    init {
        mViewPadding = mContext.resources.getDimension(R.dimen.playlist_item_selection_padding).toInt()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): RecyclerView.ViewHolder? {
        when (i) {
            BaseAdapter.TYPE_HEADER -> {
                val viewItem = LayoutInflater.from(viewGroup.context).inflate(R.layout.playlist_header, viewGroup, false)
                return ViewHolderHeader(viewItem)
            }

            BaseAdapter.TYPE_ITEM -> {
                val viewItem = LayoutInflater.from(viewGroup.context).inflate(R.layout.playlist_item_layout, viewGroup, false)
                return ViewHolderItem(viewItem)
            }

            else -> return null
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        if (mIPlaylist != null) {
            if (viewHolder is ViewHolderItem) {
                mIPlaylist!!.goToPosition((i - 1).toLong())

                // Main
                viewHolder.mImageTrack!!.setImageResource(R.drawable.image_item_track)
                viewHolder.mCountTrack!!.text = i.toString()
                viewHolder.mNameTrack!!.text = mIPlaylist!!.trackName.toString()
                viewHolder.mNameTrack!!.setTypeface(viewHolder.mNameTypeface, Typeface.NORMAL)

                // Info
                if (mIsExpand) {
                    viewHolder.mInfoLayout!!.visibility = View.VISIBLE
                    viewHolder.mDurationTrack!!.text = TimeTool.getDuration(mIPlaylist!!.trackDuration)
                    viewHolder.mArtistsTrack!!.text = mIPlaylist!!.trackArtist.toString()
                    viewHolder.mAlbumTrack!!.text = mIPlaylist!!.trackAlbum.toString()
                    viewHolder.mDateTrack!!.text = TimeTool.getDateTime(mIPlaylist!!.trackDateModified)
                } else {
                    viewHolder.mInfoLayout!!.visibility = View.GONE
                }

                // Background for current position or image_search
                if (mIPlaylist!!.trackId == CursorFactory.instance.trackId) {
                    viewHolder.itemView.setBackgroundResource(R.drawable.drawable_recycleview_item_selection)
                    viewHolder.itemView.setPadding(mViewPadding, mViewPadding, mViewPadding, mViewPadding)
                    viewHolder.mNameTrack!!.setTypeface(viewHolder.mNameTrack!!.typeface, Typeface.BOLD)
                } else if (mSearchedPosition == i.toLong()) {
                    viewHolder.itemView.setBackgroundResource(R.drawable.drawable_recycleview_item_find)
                    viewHolder.itemView.setPadding(mViewPadding, mViewPadding, mViewPadding, mViewPadding)
                } else {
                    viewHolder.itemView.setBackgroundColor(Color.TRANSPARENT)
                    viewHolder.itemView.setPadding(0, 0, 0, 0)
                }

            } else if (viewHolder is ViewHolderHeader) {
                viewHolder.mNamePlaylist!!.text = CursorTool.getPlaylistNameById(mContext.contentResolver,
                        mIPlaylist!!.playlistId)
                viewHolder.mCountTracks!!.text = totalTracks.toString()
                viewHolder.mTotalTimeTracks!!.text = TimeTool.getDuration(totalTime)
            }
        }
    }

    override fun getItemCount(): Int {
        return (if (mIPlaylist != null) mIPlaylist!!.size().toInt() else 0) + 1
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) {
            BaseAdapter.TYPE_HEADER
        } else BaseAdapter.TYPE_ITEM
    }

    override fun getItemId(position: Int): Long {
        return if (mIPlaylist != null && mIPlaylist!!.goToPosition(position.toLong())) {
            mIPlaylist!!.trackId
        } else IPlaylist.ERROR_CODE

    }

    fun setPlaylist(iPlaylist: IPlaylist) {
        mIPlaylist = iPlaylist
        notifyDataSetChanged()
    }

    fun setExpand(isExpand: Boolean) {
        mIsExpand = isExpand
        notifyDataSetChanged()
    }

    fun searchMatch(text: String): Int {
        if (mIPlaylist != null) {
            val position = mIPlaylist!!.find(mPreviousSearchPosition, text)
            mPreviousSearchPosition = position + 1
            return position.toInt()
        }

        return IPlaylist.ERROR_CODE.toInt()
    }

    fun setSearchedPosition(position: Long) {
        mSearchedPosition = position
    }

    inner class ViewHolderHeader(view: View) : RecyclerView.ViewHolder(view) {
        @BindView(R.id.playlist_header_name_value)
        lateinit var mNamePlaylist: TextView
        @BindView(R.id.playlist_header_count_value)
        lateinit var mCountTracks: TextView
        @BindView(R.id.playlist_header_total_time_value)
        lateinit var mTotalTimeTracks: TextView

        init {
            ButterKnife.bind(this, view)
        }
    }

    inner class ViewHolderItem(view: View) : RecyclerView.ViewHolder(view) {

        /*
        * Main views
        * */
        @BindView(R.id.playlist_track_image)
        lateinit var mImageTrack: ImageView
        @BindView(R.id.playlist_track_count)
        lateinit var mCountTrack: TextView
        @BindView(R.id.playlist_track_name)
        lateinit var mNameTrack: TextView

        /*
        * Optional
        * */
        @BindView(R.id.playlist_track_item_info_layout)
        lateinit var mInfoLayout: ConstraintLayout
        @BindView(R.id.playlist_track_duration_value)
        lateinit var mDurationTrack: TextView
        @BindView(R.id.playlist_track_artist_value)
        lateinit var mArtistsTrack: TextView
        @BindView(R.id.playlist_track_albums_value)
        lateinit var mAlbumTrack: TextView
        @BindView(R.id.playlist_track_date_value)
        lateinit var mDateTrack: TextView

        /*
        * Typeface
        * */
        internal var mNameTypeface: Typeface

        init {
            ButterKnife.bind(this, view)
            view.setOnClickListener {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(view, layoutPosition - 1)
                }
            }

            mNameTypeface = mNameTrack!!.typeface
        }
    }

}
